tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/UninterpretedLiteralCacheTest.scala

```diff
@@ -0,0 +1,106 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux
+
+import at.forsyte.apalache.tla.bmcmt.PureArena
+import at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches.UninterpretedLiteralCache
+import at.forsyte.apalache.tla.lir.StrT1
+import at.forsyte.apalache.tla.types.{tla, ModelValueHandler}
+import org.junit.runner.RunWith
+import org.scalatest.BeforeAndAfterEach
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalatestplus.junit.JUnitRunner
+
+@RunWith(classOf[JUnitRunner])
+class UninterpretedLiteralCacheTest extends AnyFunSuite with BeforeAndAfterEach {
+
+  var cache: UninterpretedLiteralCache = new UninterpretedLiteralCache
+
+  def tpAndIdx(s: String): (String, String) = {
+    val (utype, idx) = ModelValueHandler.typeAndIndex(s).getOrElse((StrT1, s))
+    (utype.toString, idx)
+  }
+
+  override def beforeEach(): Unit = {
+    cache = new UninterpretedLiteralCache
+  }
+
+  test("Cache returns stored values after the first call to getOrCreate") {
+    val str: String = "idx"
+
+    val utypeAndIdx = tpAndIdx(str)
+
+    val arena = PureArena.empty
+
+    // No cached value for the pair
+    assert(cache.get(utypeAndIdx).isEmpty)
+
+    val (newArena, iCell) = cache.getOrCreate(arena, utypeAndIdx)
+
+    // pair now cached, arena has changed
+    assert(cache.get(utypeAndIdx).nonEmpty && newArena != arena)
+
+    val (newArena2, iCell2) = cache.getOrCreate(newArena, utypeAndIdx)
+
+    // 2nd call returns the _same_ arena and the previously computed cell
+    assert(newArena == newArena2 && iCell == iCell2)
+  }
+
+  test("Same index of different types is cached separately") {
+    val str1: String = "idx"
+    val str2: String = "idx_OF_A"
+    val str3: String = "idx_OF_B"
+
+    val pa1 = tpAndIdx(str1)
+    val pa2 = tpAndIdx(str2)
+    val pa3 = tpAndIdx(str3)
+
+    val arena = PureArena.empty
+
+    val (newArena1, cell1) = cache.getOrCreate(arena, pa1)
+
+    assert(arena != newArena1)
+
+    val (newArena2, cell2) = cache.getOrCreate(newArena1, pa2)
+
+    assert(newArena2 != newArena1 && cell2 != cell1)
+
+    val (newArena3, cell3) = cache.getOrCreate(newArena2, pa3)
+
+    assert(newArena3 != newArena2 && cell3 != cell2)
+  }
+
+  test("Constraints are only added when addConstraintsForElem is explicitly called, and only once per value") {
```

Comment:
Good catch, slipped my mind. I am a little torn though, given how impractical calling distinct on single elements is, I should probably try and find a way to discourage/disable invoking the "-ForElem" method.

Reply:
Funnily enough, adding these tests helped find a logic bug: In `ModelValueCache`, the line [1]: https://github.com/informalsystems/apalache/blob/4e0094ee5d03a068188a85dcdc8ead021b17033e/tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/caches/ModelValueCache.scala#L30
In `create` returns a collection of all other cells of the same type as the newest cell. However, in `AbstractCache` https://github.com/informalsystems/apalache/blob/4e0094ee5d03a068188a85dcdc8ead021b17033e/tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/caches/AbstractCache.scala#L52-L60
performs `create` _before_ `addToCache`. If one would have used line [1] as is in `addConstaintsForElem`, this would create UNSAT constraints, because that method gets called _after_ both `create` _and_ `addToCache`, effectively creating `distinct(x, ..., x, ...)` constraints, since the latest cell is now part of the `others` collection too.
Note that the `addAllConstraints` behavior is unaffected by this, so the "normal" use case wouldn't have been affected, but I'm sure we'd be dealing with the single-elem constraints bug down the line.

Reply:
Good find!